compiler/src/dotty/tools/dotc/semanticdb/PPrint.scala

```diff
@@ -0,0 +1,273 @@
+            sb.append(s"{ ${selfStr} ${declsStr} }")
+          }
+          sb.toString
+        case MethodSignature(tparams, paramss, res) =>
+          val sb = new StringBuilder()
+          if (tparams.infos.nonEmpty)
+            sb.append(tparams.infos.map(pprintDef).mkString("[", ", ", "]"))
+          paramss.foreach { params =>
+            val paramsStr = params.infos.map(pprintDef).mkString("(", ", ", ")")
+            sb.append(paramsStr)
+          }
+          sb.append(s": ${pprint(res)}")
+          sb.toString
+        case TypeSignature(tparams, lo, hi) =>
+          val sb = new StringBuilder()
+          if (tparams.infos.nonEmpty)
+            sb.append(tparams.infos.map(pprintDef).mkString("[", ", ", "]"))
+          if (lo == hi) {
+            sb.append(s" = ${pprint(lo)}")
+          } else {
+            lo match
+              case TypeRef(Type.Empty, "scala/Nothing#", Nil) => ()
+              case lo => sb.append(s" >: ${pprint(lo)}")
+            hi match
+              case TypeRef(Type.Empty, "scala/Any#", Nil) => ()
+              case TypeRef(Type.Empty, "java/lang/Object#", Nil) => ()
+              case hi => s" <: ${pprint(hi)}"
+          }
+          sb.toString
+        case ValueSignature(tpe) =>
+          pprint(tpe)
+        case _ =>
+          "<?>"
+
+    private def pprint(tpe: Type): String = {
+      def prefix(tpe: Type): String = tpe match
+        case TypeRef(pre, sym, args) =>
+          val preStr = pre match {
+            case _: SingleType | _: ThisType | _: SuperType =>
+              s"${prefix(pre)}."
+            case Type.Empty => ""
+            case _ =>
+              s"${prefix(pre)}#"
+          }
+          val argsStr = if (args.nonEmpty) args.map(normal).mkString("[", ", ", "]") else ""
+          s"${preStr}${pprintRef(sym)}${argsStr}"
+        case SingleType(pre, sym) =>
+          pre match {
+            case Type.Empty => pprintRef(sym)
+            case _ =>
+              s"${prefix(pre)}.${pprintRef(sym)}"
+          }
+        case ThisType(sym) =>
+          s"${pprintRef(sym)}.this"
+        case SuperType(pre, sym) =>
+          s"${prefix(pre)}.super[${pprintRef(sym)}]"
+        case ConstantType(const) =>
+          pprint(const)
+        case IntersectionType(types) =>
+          types.map(normal).mkString(" & ")
+        case UnionType(types) =>
+          types.map(normal).mkString(" | ")
+        case WithType(types) =>
+          types.map(normal).mkString(" with ")
+        case StructuralType(utpe, decls) =>
+          val declsStr =
+            if (decls.infos.nonEmpty)
+              s"{ ${decls.infos.map(pprintDef).mkString("; ")} }"
+            else "{}"
+          s"${normal(utpe)} ${declsStr}"
+        case AnnotatedType(anns, utpe) =>
+          s"${normal(utpe)} ${anns.map(pprint).mkString(" ")}"
+        case ExistentialType(utpe, decls) =>
+          val sdecls = decls.infos.map(_.displayName).mkString("; ")
```

Comment:
here you should first call `notes.enter` on each of the `decls`, then they will print correctly when referenced within `utpe`. It would be nice to also print the bounds of each decl, not just the `displayName`

Reply:
done:+1: https://github.com/lampepfl/dotty/pull/12885/commits/1622980bc234e6b5aa69ea04c69502d89fdf66c9